model/host/host.go

```diff
@@ -301,6 +301,36 @@ func (h *Host) SetStatus(status, user string, logs string) error {
 	)
 }
 
+// SetStatusAtomically is the same as SetStatus but only updates the host if its
+// status in the database matches currentStatus.
+func (h *Host) SetStatusAtomically(newStatus, currentStatus, user string, logs string) error {
```

Comment:
are there cases when we wouldn't want this behavior? feels like this should be the main SetStatus method and we should have an "OverrideSetStatus" for anything where we want the bad behavior. 

also shouldn't we be able to see "currentStatus" in h.Status rather than as an arg?

Reply:
I changed SetStatusAtomically to use the current host status. The default will still be to unconditionally set the host status since there are many jobs that set the status unconditionally and some may fail without the ability to retry.